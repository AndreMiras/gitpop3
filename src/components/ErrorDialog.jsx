import React, { useState } from 'react';
import PropTypes from 'prop-types';
import { Modal } from 'react-bootstrap';

const ErrorDialog = ({ detail, onClose }) => {
  const [show, setShow] = useState(true);
  const handleClose = () => {
    setShow(false);
    onClose();
  };

  // note animation is disabled to workaround an upstream issue:
  // https://github.com/react-bootstrap/react-bootstrap/issues/5075
  return (
    <Modal show={show} onHide={handleClose} animation={false}>
      <Modal.Header closeButton className="bg-warning">
        <Modal.Title>Error</Modal.Title>
      </Modal.Header>
      <Modal.Body>{detail}</Modal.Body>
    </Modal>
  );
};
ErrorDialog.propTypes = {
  detail: PropTypes.string.isRequired,
  onClose: PropTypes.func,
};
ErrorDialog.defaultProps = {
  onClose: () => null,
};

export default ErrorDialog;
