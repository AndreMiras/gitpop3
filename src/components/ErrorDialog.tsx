import { FunctionComponent, useState } from "react";
import { Modal } from "react-bootstrap";

type ErrorDialogProps = {
  detail: string;
  onClose?: () => void;
};

const ErrorDialog: FunctionComponent<ErrorDialogProps> = ({
  detail,
  onClose,
}) => {
  const [show, setShow] = useState(true);
  const handleClose = () => {
    setShow(false);
    onClose && onClose();
  };

  // note animation is disabled to workaround an upstream issue:
  // https://github.com/react-bootstrap/react-bootstrap/issues/5075
  return (
    <Modal show={show} onHide={handleClose} animation={false}>
      <Modal.Header closeButton className="bg-warning">
        <Modal.Title>Error</Modal.Title>
      </Modal.Header>
      <Modal.Body>{detail}</Modal.Body>
    </Modal>
  );
};

export default ErrorDialog;
