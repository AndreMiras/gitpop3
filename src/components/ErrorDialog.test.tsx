import React, { ReactNode, ReactPortal } from "react";
import ReactDOM from "react-dom";
import renderer from "react-test-renderer";
import { fireEvent, render, screen } from "@testing-library/react";
import ErrorDialog from "./ErrorDialog";

describe("ErrorDialog", () => {
  /**
   * Workaround for the modal not rendering properly in tests.
   * https://stackoverflow.com/a/58076920/185510
   */
  const oldCreatePortal = ReactDOM.createPortal;
  beforeAll(() => {
    ReactDOM.createPortal = (node: ReactNode): ReactPortal =>
      node as ReactPortal;
  });

  afterAll(() => {
    ReactDOM.createPortal = oldCreatePortal;
  });

  test("renders", () => {
    const onClose = () => ({});
    const tree = renderer
      .create(<ErrorDialog detail="Error details renders" onClose={onClose} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("onClose omitted", () => {
    const tree = renderer
      .create(<ErrorDialog detail="Error details onClose omitted" />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  test("onClose", (done) => {
    render(<ErrorDialog detail="Error details onClose" onClose={done} />);
    const closeButton = screen.getByRole("button", {
      hidden: true,
    });
    fireEvent.click(closeButton);
  });
});
