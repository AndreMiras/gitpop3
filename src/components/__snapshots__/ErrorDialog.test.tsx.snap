// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ErrorDialog onClose omitted 1`] = `
Array [
  <div
    className="modal-backdrop show"
    onClick={[Function]}
  />,
  <div
    aria-modal={true}
    className="modal"
    onClick={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={
      Object {
        "display": "block",
      }
    }
    tabIndex={-1}
  >
    <div
      className="modal-dialog"
      onMouseDown={[Function]}
    >
      <div
        className="modal-content"
      >
        <div
          className="bg-warning modal-header"
        >
          <div
            className="modal-title h4"
          >
            Error
          </div>
          <button
            className="close"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
            <span
              className="sr-only"
            >
              Close
            </span>
          </button>
        </div>
        <div
          className="modal-body"
        >
          Error details onClose omitted
        </div>
      </div>
    </div>
  </div>,
]
`;

exports[`ErrorDialog renders 1`] = `
Array [
  <div
    className="modal-backdrop show"
    onClick={[Function]}
  />,
  <div
    aria-modal={true}
    className="modal"
    onClick={[Function]}
    onMouseUp={[Function]}
    role="dialog"
    style={
      Object {
        "display": "block",
      }
    }
    tabIndex={-1}
  >
    <div
      className="modal-dialog"
      onMouseDown={[Function]}
    >
      <div
        className="modal-content"
      >
        <div
          className="bg-warning modal-header"
        >
          <div
            className="modal-title h4"
          >
            Error
          </div>
          <button
            className="close"
            onClick={[Function]}
            type="button"
          >
            <span
              aria-hidden="true"
            >
              ×
            </span>
            <span
              className="sr-only"
            >
              Close
            </span>
          </button>
        </div>
        <div
          className="modal-body"
        >
          Error details renders
        </div>
      </div>
    </div>
  </div>,
]
`;
